File: scala/src/com/nixprime/nijikawa/Nijikawa.scala
package com.nixprime.nijikawa

import dram.SimpleDram
import mem_trace.{UsimmTraceReader, MemTraceCore}

object Nijikawa {
  val terminationCycle: Long = 100000000 // 100 million

  def main(args: Array[String]) {
    val sim = new Simulator
    val dram = new SimpleDram(sim)
    val core_trace_reader = new UsimmTraceReader("/home/jamiel/src/arch_sims/usimm/input/comm1")
    val core = new MemTraceCore(sim, core_trace_reader, dram.receiveRequest)

    core.superscalarWidth = 4
    core.robSize = 192

    core.init()
    dram.init()

    while (sim.now < terminationCycle) {
      core.tick()
      dram.tick()
      sim.now += 1
    }

    println(core.insnsRetired.toString + " instructions retired in " + terminationCycle.toString +
            " cycles")
  }
}

File: rust/nijikawa.rs
extern mod extra;

use core::*;
use dram::*;
use mem_request::*;
use simulator::*;
use trace::*;
use usimm_trace_reader::*;

pub mod core;
pub mod dram;
pub mod mem_request;
pub mod simulator;
pub mod trace;
pub mod usimm_trace_reader;
pub mod util;

static sim_cycles: i64 = 100000000;
static trace_filename: &'static str = "/home/jamiel/src/usimm-v1.3/input/comm1";

fn main() {
    let mut sim = Simulator::new();
    let mut trace_reader = UsimmTraceReader::new(trace_filename);
    let mut dram = Dram::new(&sim, 1, 4);
    let mut core = @mut Core::new(&sim, &mut trace_reader as &mut TraceReader,
                                  &mut dram as &mut MemRequestReceiver, 4, 192);
    while sim.now() < sim_cycles {
        core.tick();
        dram.tick();
        sim.tick();
    }
    println!("{:u} instructions retired in {:i} cycles", core.insns_retired,
             sim.now())
}
